### PR TITLE
Add EtherNet/IP and CIP I/O protocols

### PR DESCRIPTION
**Pull request by jacobvw**

### CODE REVIEW

File: lib/tcp/lpi_ethernetip.cc [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	return false;
+}
+
+static inline bool match_ethernetip(lpi_data_t *data, lpi_module_t *mod UNUSED) {
+
+	if (data->payload_len[0] < 24 || data->payload_len[1] < 24)
+		return false;
+
+	if (data->server_port != 44818)
```
Comment (salcock) [inline lib/tcp/lpi_ethernetip.cc]: This should return false if neither server nor client port is 44818 -- we can't trust libflowmanager to get the server-client ordering right ;)

File: lib/udp/lpi_cip_io.cc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	if (MATCH(data->payload[1], 0x02, 0x00, 0x02, 0x80) &&
+            MATCH(data->payload[0], 0x00, 0x00, 0x00, 0x00))
+		return true;
+
+	return false;
+}
+
+static inline bool match_cip_io(lpi_data_t *data, lpi_module_t *mod UNUSED) {
+
+	if (data->server_port != 2222 && data->client_port != 2222)
```
Comment (salcock) [inline lib/udp/lpi_cip_io.cc]: Should this be an `||` instead? Are there cases where only one endpoint is using port 2222?

File: lib/udp/lpi_cip_io.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	if (match_cip(data))
+		return true;
+
+	return false;
+}
+
+static lpi_module_t lpi_cip_io = {
+	LPI_PROTO_UDP_CIP_IO,
+	LPI_CATEGORY_ICS,
+	"CIP I/O",
```
Comment (salcock) [inline lib/udp/lpi_cip_io.cc]: Replace whitespace in protocol string with `_` -- many of our tools produce space-separated output and are going to break if we have names with spaces in them

File: lib/udp/lpi_ethernetip.cc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            match_command(data->payload[1], data->payload_len[1]))
+		return true;
+
+	return false;
+}
+
+static lpi_module_t lpi_ethernetip_udp = {
+	LPI_PROTO_UDP_ETHERNETIP,
+	LPI_CATEGORY_ICS,
+	"EtherNet/IP UDP",
```
Comment (salcock) [inline lib/udp/lpi_ethernetip.cc]: Another protocol name with a space to fix.

File: lib/udp/lpi_ethernetip.cc [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	2,
```
Comment (salcock) [inline lib/udp/lpi_ethernetip.cc]: Maybe increase this value to 100 (or higher) -- false positives seem a bit likely and I'd rather give other rules a chance to match first.

**Review by salcock — CHANGES_REQUESTED** [review]

File: lib/udp/lpi_cip_io.cc (hunk 2 above)
Comment (jacobvw) [reply lib/udp/lpi_cip_io.cc]: All the captures I have seen use port 2222 for client and server.

Comment (salcock) [reply lib/udp/lpi_cip_io.cc]: Ok, make this an `||` then -- if either port is not 2222 then you probably want the rule to return false for now.

If you do start seeing cases where only one endpoint is using port 2222 then we can revert this then.

Comment (jacobvw) [reply lib/udp/lpi_cip_io.cc]: done :)

**Pull request merged**